### PR TITLE
wolfssl: align CA_FALLBACK support with OpenSSL/GnuTLS providers

### PR DESCRIPTION
**Pull request by michael-o**
Here is how I have tested:
```
CONFIGURE="./configure --prefix=/tmp/local --with-wolfssl=/usr/local"
CURL="/tmp/local/bin/curl --verbose -o /dev/null -s"
EXTERNAL_URL="https://google.com"
INTERNAL_URL="https://deblndw011x.ad001.siemens.net"

$CONFIGURE

$CURL $EXTERNAL_URL # 1-1.txt
$CURL $INTERNAL_URL # 1-2.txt
$CURL --cacert /tmp/cacerts.crt $INTERNAL_URL # 1-3.txt

$CONFIGURE --without-ca-bundle --without-ca-path

$CURL $EXTERNAL_URL # 2-1.txt
$CURL $INTERNAL_URL # 2-2.txt
$CURL --capath /etc/ssl/certs $EXTERNAL_URL # 2-3.txt
$CURL --capath /usr/share/certs/trusted $INTERNAL_URL # 2-4.txt
$CURL --capath /etc/ssl/certs $INTERNAL_URL # 2-5.txt

$CONFIGURE --without-ca-bundle --without-ca-path --with-ca-fallback

$CURL $EXTERNAL_URL # 3-1.txt
$CURL $INTERNAL_URL # 3-2.txt

```
All tests performed with wolfSSL 5.6.0 on FreeBSD 12.4-STABLE amd64.

The idea behind the test appraoch is to have the Mozilla NSS CA cert bundle and the enterprise internal CA certs used against the external and internal target.

[1-1.txt](https://github.com/curl/curl/files/12714385/1-1.txt)
[1-2.txt](https://github.com/curl/curl/files/12714384/1-2.txt)
[1-3.txt](https://github.com/curl/curl/files/12714383/1-3.txt)

[2-1.txt](https://github.com/curl/curl/files/12714382/2-1.txt)
[2-2.txt](https://github.com/curl/curl/files/12714381/2-2.txt)
[2-3.txt](https://github.com/curl/curl/files/12714380/2-3.txt)
[2-4.txt](https://github.com/curl/curl/files/12714376/2-4.txt)
[2-5.txt](https://github.com/curl/curl/files/12714379/2-5.txt)

[3-1.txt](https://github.com/curl/curl/files/12714378/3-1.txt)
[3-2.txt](https://github.com/curl/curl/files/12714377/3-2.txt)

### CODE REVIEW

**Comment by jay** [comment]
> * Swap CA native for CA_FALLBACK as done with OpenSSL and GnuTLS: Especially the behavior compared to GnuTLS was inconsistent because wolfSSL_CTX_load_system_CA_certs() behaves internally like gnutls_certificate_set_x509_system_trust(), means loading from Windows native store and macOS native store already and completely opaque to libcurl. OpenSSL does not provide such a feature, therefore it makes sense to support '--ca-native' only on OpenSSL.

I'm not following this. We have [CURLSSLOPT_NATIVE_CA](https://curl.se/libcurl/c/CURLOPT_SSL_OPTIONS.html) that will load the native CA store for wolfSSL. You seem to be saying that the ca fallback build option should do that instead. Why not both?

**Comment by michael-o** [comment]
> > * Swap CA native for CA_FALLBACK as done with OpenSSL and GnuTLS: Especially the behavior compared to GnuTLS was inconsistent because wolfSSL_CTX_load_system_CA_certs() behaves internally like gnutls_certificate_set_x509_system_trust(), means loading from Windows native store and macOS native store already and completely opaque to libcurl. OpenSSL does not provide such a feature, therefore it makes sense to support '--ca-native' only on OpenSSL.
> 
> I'm not following this. We have [CURLSSLOPT_NATIVE_CA](https://curl.se/libcurl/c/CURLOPT_SSL_OPTIONS.html) that will load the native CA store for wolfSSL. You seem to be saying that the ca fallback build option should do that instead. Why not both?

Not quite, I say both can, hence my note in the PR. How or where from wolfSSL loads the default truststore is opaque to us, whether it is truly the native one, we don't know from the Doxygen description, but if you look at how wolfSSL is implemented then you will see that it is identical with the GnuTLS implementation (not in libcurl, the lib itself), thus I would expect that wolfSSL does not require `--ca-native` if one has the fallback configured. Either way they are inconsitent this is why I have chosen to align wolfSSL to the current behavior of the other providers. Does this work for you?

**Comment by bagder** [comment]
> therefore it makes sense to support '--ca-native' only on OpenSSL.

I strongly disagree. Remember that `configure --with-ca-fallback` is the special case, it is not the default build. The API offers users to select to use the native CA, it is not a built in forced choice.

**Comment by michael-o** [comment]
> > therefore it makes sense to support '--ca-native' only on OpenSSL.
> 
> I strongly disagree. Remember that `configure --with-ca-fallback` is the special case, it is not the default build. The API offers users to select to use the native CA, it is not a built-in forced choice.

I don't disagree with you, but as said then based on your logic we need to introduce --ca-native with GnuTLS as well to make it consistent. So from my understanding is what needs to be done:
* retain `--ca-native` in wolfSSL
* introduce `--ca-native` in GnuTLS like in wolfSSL

then all of the three providers have identical behavior for both `ca-native` and `CA_FALLBACK`. Does this work now?

The GnuTLS change needs a separate PR, though.

**Comment by michael-o** [comment]
Working on an update...

**Comment by michael-o** [comment]
@jay @bagder I think this is now what you expect to see...

File: docs/libcurl/opts/CURLOPT_CAPATH.3 [hunk 1]
```diff
@@ -60,8 +60,8 @@ if(curl) {
 }
 .fi
 .SH AVAILABILITY
-This option is supported by the OpenSSL, GnuTLS and mbedTLS (since 7.56.0)
-backends.
+This option is supported by the OpenSSL, GnuTLS, mbedTLS (since 7.56.0) and
+wolfSSL (since 8.4.0) backends.
```
Comment (bagder) [inline docs/libcurl/opts/CURLOPT_CAPATH.3:64]: I don't think this is correct. The code already uses this argument before this change. Since at least 7.65.2 (deb9462ff)

Comment (bagder) [reply docs/libcurl/opts/CURLOPT_CAPATH.3:64]: I think the feature check came later, which then made it not work again - unless set in configure.

Comment (michael-o) [reply docs/libcurl/opts/CURLOPT_CAPATH.3:64]: Right, but it was not exposed through CLI, so it was incomplete. I will drop this hunk. Good catch.

Comment (michael-o) [reply docs/libcurl/opts/CURLOPT_CAPATH.3:64]: Do you want to address this doc fix in your PR as well?

Comment (michael-o) [reply docs/libcurl/opts/CURLOPT_CAPATH.3]: I have dropped this hunk.

**Comment by bagder** [comment]
> we need to introduce `--ca-native` with GnuTLS

Yes!

**Comment by michael-o** [comment]
> > we need to introduce `--ca-native` with GnuTLS
> 
> Yes!

Right, then let's wait until these two related PRs are settled and I can work on the GnuTLS PR.

**Comment by michael-o** [comment]
@bagder Rebased and reduced to a minimum. Please have a look.

**Comment by bagder** [comment]
This I believe alter the meaning of the original `--ca-fallback` implementation slightly, as this no longer sets the "TLS library's own default". Without any documentation explaining it.

What's the need for this option to do this?

**Comment by michael-o** [comment]
> This I believe alter the meaning of the original `--ca-fallback` implementation slightly, as this no longer sets the "TLS library's own default". Without any documentation explaining it.
> 
> What's the need for this option to do this?

I don't understand your statement. This requests wolfssl at compile time to use default like OpenSSL. Can you rephrase please?

**Comment by michael-o** [comment]
> This I believe alter the meaning of the original `--ca-fallback` implementation slightly, as this no longer sets the "TLS library's own default". Without any documentation explaining it.
> 
> What's the need for this option to do this?

I re-read your statement and think that it is actually the opposite. The default one should be always used unless explicitly overwritten/disabled by the users. I think that fallback is an unfortunate name. It should have been `--ca-default` would have been better.

**Comment by jay** [comment]
> This requests wolfssl at compile time to use default like OpenSSL. Can you rephrase please?

> The default one should be always used unless explicitly overwritten/disabled by the users.

I think what @bagder is getting at is it's not the default? OpenSSL has hard coded path names so when you call X509_STORE_set_default_paths it actually loads certificates from those default paths.

**Comment by michael-o** [comment]
> > This requests wolfssl at compile time to use default like OpenSSL. Can you rephrase please?
> 
> > The default one should be always used unless explicitly overwritten/disabled by the users.
> 
> I think what @bagder is getting at is it's not the default? OpenSSL has hard coded path names so when you call X509_STORE_set_default_paths it actually loads certificates from those default paths.

So does wolfSSL try to load from static paths of you look into the source code. I see no difference to OpenSSL here.

**Comment by bagder** [comment]
> So does wolfSSL try to load from static paths of you look into the source code. I see no difference to OpenSSL here.

But loading the system certs is not default for either.

**Comment by michael-o** [comment]
> > So does wolfSSL try to load from static paths of you look into the source code. I see no difference to OpenSSL here.
> 
> But loading the system certs is not default for either.

No, that is why is a compile time option for all three providers. Where and now the CA certs reside on a system is an implementation detail and that is solved by the crypto framework. GnuTLS in libcurl *already* exposes the behavior the patch introduces for wolfSSL.

**Comment by michael-o** [comment]
> `--with-ca-fallback Use the built in CA store of the SSL library` but as you pointed out that's not what we're doing with GnuTLS.
> 
> https://github.com/curl/curl/blob/6fa1d817e5b1a00d7d0c8168091877476b499317/lib/vtls/gtls.c#L500-L506
> 
> My guess is that `--with-ca-fallback` was written with OpenSSL in mind since it may have its own "built in CA store", ie something separate from the packaged certificate location?

This is one of the reason reasons why I said that it should be `--with-ca-default` whatever default means is a detail of the crypto framework. In contrast to OpenSSL both other providers don't have a notion of a conf dir, as far as I understand, but that does not mean that cannot use the cert store which is designed for OpenSSL, infact they do query well-known locations were OpenSSL places hashed subject links. I don't see anything harmful to have this option especially it makes using wolfSSL as smooth as the other providers.

**Comment by bagder** [comment]
> This is one of the reason reasons why I said that it should be --with-ca-default

But you are currently advocating a change of `--with-ca-fallback` and this option was created and has been used for a specific purpose. I feel that this change changes that original meaning and purpose for this option.

I don't think the "fallback" means use the system CV store. The fact that it does for GnuTLS is a bug.

> I don't see anything harmful to have this option especially it makes using wolfSSL as smooth as the other providers.

I find it already being that! Without a build option that makes it use the system CA store without me asking for it.

**Comment by michael-o** [comment]
> > This is one of the reason reasons why I said that it should be --with-ca-default
> 
> But you are currently advocating a change of `--with-ca-fallback` and this option was created and has been used for a specific purpose. I feel that this change changes that original meaning and purpose for this option.

What is the specific purpose? It lacks obviously documentation, no? I am trying to use what is there and makes sense, and not reinvent the wheel if not necessary here.

> I don't think the "fallback" means use the system CA store. The fact that it does for GnuTLS is a bug.

Granted, that maybe true. Lack of a specific could mean either.

> > I don't see anything harmful to have this option especially it makes using wolfSSL as smooth as the other providers.
> 
> I find it already being that! Without a build option that makes it use the system CA store without me asking for it.

OK, that would mean the following:

* Deprecation of `--with-ca-fallback`
* Introduction of `--with-ca-default` which keeps where and how to load CA certificates at the descretion of the framework as long as it provides a function to load them.
* The new build time option would apply to OpenSSL/GnuTLS and wolfSSL for now.

This PR would change to:
```
#ifdef CURL_CA_DEFAULT
...
#endif
```

Subsequent PRs will follow.


WDYT?

**Comment by bagder** [comment]
> Introduction of --with-ca-default

Why? We provide this ability in the API, why do we need a build work-around to enable this by default?

**Comment by michael-o** [comment]
> > Introduction of --with-ca-default
> 
> Why? We provide this ability in the API, why do we need a build work-around to enable this by default?

This is what I understood from your objections on the abuse of CA_FALLBACK.

**Comment by michael-o** [comment]
> > This is what I understood from your objections on the abuse of CA_FALLBACK.
> 
> I objected to that, yes.
> 
> Now you're proposing another build option with another functionality and I'm asking why we want this.

Please propose what you expect to see here, since I don't understand it.

**Comment by bagder** [comment]
> Please propose what you expect to see here, since I don't understand it.

*You* propose this PR. This is a change *you* propose. I want you to explain to me *why* we should merge it. What's the motivation for this change and what is the use case?

I don't expect it at all. I don't see a need for it.

**Comment by michael-o** [comment]
> > Please propose what you expect to see here, since I don't understand it.
> 
> _You_ propose this PR. This is a change _you_ propose. I want you to explain to me _why_ we should merge it. What's the motivation for this change and what is the use case?
> 
> I don't expect it at all. I don't see a need for it.

I did already many times: consistency with other providers.  Even if you don't see it, it exists for others.

**Comment by bagder** [comment]
I don't see how this helps with consistency for users. I think it rather adds inconsistencies and more build combinations to keep track of.

**Comment by michael-o** [comment]
> I don't see how this helps with consistency for users. I think it rather adds inconsistencies and more build combinations to keep track of.

Because it then works out of the box with more providers with the same switch...no new switch necessary.

**Comment by bagder** [comment]
> Because it then works out of the box with more providers with the same switch...no new switch necessary.

You are *arguing* for a new switch.

**Comment by bagder** [comment]
I don't think this discussion is going forward much. I don't think you have presented compelling arguments for merging this PR. I don't think this is a good idea.

**Comment by michael-o** [comment]
> I don't think this discussion is going forward much. I don't think you have presented compelling arguments for merging this PR. I don't think this is a good idea.

I have proposed a simple opt-in, seven-line improvement and *you* made rocket science out of it by going circles.

**Comment by bagder** [comment]
I'm sorry but:
1. I care about details, even if you think they are small and insignificant
2. I don't want  to add more confusion but rather work in the opposite direction
3. new build options should be avoided unless properly motivated and thought through
4. History tells me that I am usually the person who will be stuck with the code 10 years later
